perfkitbenchmarker/background_tasks.py

```diff
@@ -0,0 +1,578 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Background tasks that propagate PKB thread context.
+
+PKB tries its best to clean up provisioned resources upon SIGINT. By default,
+Python raises a KeyboardInterrupt upon a SIGINT, but none of the built-in
+threading module classes are designed to handle a KeyboardInterrupt very well:
+
+- threading.Lock has an atomic acquire method that cannot be interrupted and
+  hangs forever if the same thread tries to acquire twice. Its release method
+  can be called by any thread but raises thread.error if an unacquired Lock is
+  released.
+
+- More complicated classes (threading.RLock, threading.Event, threading.Thread,
+  Queue.Queue) use internal Locks in such a way that a KeyboardInterrupt can
+  cause a thread that has acquired a Lock to jump out of its current action
+  without releasing the Lock. For example, in the below code, a
+  KeyboardInterrupt can be raised immediately after the acquire call but before
+  entering the try block:
+    lock.acquire()
+    try:
+      ...
+    except:
+      lock.release()
+
+Taken together, this means that there is a possibility to leave an internal Lock
+acquired, and when later cleanup steps on the same or different thread attempt
+to acquire the Lock, they will hang forever, unresponsive to even a second
+KeyboardInterrupt. A KeyboardInterrupt during Thread.start() or Thread.join()
+can even trigger an unbalanced acquire on a global lock used to keep track of
+active threads, so that later attempts to start or join any Thread will hang
+forever.
+
+While it would take a significant and impractical redesign of PKB's code to
+completely eliminate any risk of deadlock following a KeyboardInterrupt, the
+code in this module is designed to allow interrupting parallel tasks while
+keeping the risk of deadlock low.
+"""
+
+import abc
+from collections import deque
+from concurrent import futures
+import ctypes
+import functools
+import logging
+import os
+import Queue
+import threading
+import time
+import traceback
+
+from perfkitbenchmarker import context
+from perfkitbenchmarker import errors
+from perfkitbenchmarker import log_util
+
+
+# For situations where an interruptable wait is necessary, a loop of waits with
+# long timeouts is used instead. This is because some of Python's built-in wait
+# methods are non-interruptable without a timeout.
+_LONG_TIMEOUT = 1000.
+
+# Constants used for polling waits. See _WaitForCondition.
+_WAIT_MIN_RECHECK_DELAY = 0.001  # 1 ms
+_WAIT_MAX_RECHECK_DELAY = 0.050  # 50 ms
+
+
+def _GetCallString(target_arg_tuple):
```

Comment:
Looks like you haven't deleted the copy of this function in vm_util - I couldn't find any references to the old function - do you still need it or can it be deleted?